ostd/src/device/dispatcher/io_mem.rs

```diff
@@ -0,0 +1,234 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Io Memory access dispatcher.
+
+use alloc::{borrow::ToOwned, vec::Vec};
+use core::ops::Range;
+
+use log::{debug, info};
+use spin::Once;
+
+use crate::{
+    device::io_mem::IoMem,
+    mm::{CachePolicy, PageFlags},
+    util::vaddr_alloc::VirtAddrAllocator,
+};
+
+/// I/O memory dispatcher that allocates I/O access to devices via MMIO.
+pub struct IoMemDispatcher {
+    allocators: Once<Vec<VirtAddrAllocator>>,
+}
+
+impl IoMemDispatcher {
+    /// Get the IO memory access with `range`. If the range is not available, then the return value will be None.
+    pub fn get(&self, range: Range<usize>) -> Option<IoMem> {
+        let mut result = None;
+        for allocator in self.allocators.get().unwrap().iter() {
+            let allocator_range = allocator.fullrange();
+            if allocator_range.start >= range.end || allocator_range.end <= range.start {
+                continue;
+            }
+
+            let start_addr = range.start;
+            let end_addr = range.end;
+            debug!("Allocating MMIO range:{:x?}..{:x?}", start_addr, end_addr);
+            if allocator.alloc_specific(range).is_err() {
+                return None;
+            }
+
+            // SAFETY: The created IoMem is guaranteed not to access physical memory or system device I/O
+            unsafe {
+                result = Some(IoMem::new(
+                    start_addr..end_addr,
+                    PageFlags::RW,
+                    CachePolicy::Uncacheable,
+                ))
+            }
+            break;
+        }
+        result
+    }
+
+    /// Remove access to a specific memory IO range.  
+    ///
+    /// All drivers in OSTD must use this method to prevent peripheral drivers from accessing illegal memory IO range.
+    pub(crate) fn remove(&self, range: Range<usize>) {
```

Comment:
I refactored the build process for the `IoMemDispatcher`. The system now first builds the `IoMemDispatcherBuilder` during the `init_on_bsp` phase according to the memory distribution. At this point, the static `IoMemDispatcher` is not yet initialized, so the injected code cannot allocate any `IoMem`. 

When the memory I/O for system devices is removed by the drivers, the `IoMemDispatcher` is then constructed and registered in the static variable.

The guarantees are as follows:

1. The allocators within the `IoMemDispatcherBuilder` are guaranteed not to allocate physical memory access.
2. Users must ensure that all system device drivers have removed their memory I/O access before registering the static `IoMemDispatcher`.

I think these guarantees are soundness based on the system devices are not hotplugging devices (IOMMU, APIC...).

Also, I think `IoMemDispatcherBuilder` is too long and a bit weird to call this name since I don't provide the `build` function.

Reply:
I suggest we can categorize or abstract the `IoMemDispatcherBuilder` to rename it like `DispatcherBuilder` when implementing the `IoPortDispatcher`.